### PR TITLE
fix: JSObjects | Sync methods | Not run consistently when Page is cloned

### LINKED ISSUE #20814: [Bug]: JSObjects | Sync methods | Not run consistently when Page is cloned
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

https://user-images.githubusercontent.com/91450662/220246222-e7e51a81-c78a-4bea-af88-ee39e89562ef.mov

### Steps To Reproduce

1. D&D a Input Widget & map it to JSObject sync method
2. Clone the Page
3. **Observe JS sync method is run on alternate times**
4. **Expected: JS Object sync method should run all the time successfully**

### Public Sample App

_No response_

### Issue video log

_No response_

### Version

Cloud/Release  v1.9.8

### PR DESCRIPTION
**Pull request by Druthi**
## Description
fix: JSObjects | Sync methods | Not run consistently when Page is cloned

We used to not wait for all the calls to fetch the entities to run evaluations however we changed this with some performance improvements. This seems to have been missed for when we clone a page.

Fixes #20814 

## Type of change
- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?
- Manual
- Cypress

### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

### QA activity:
- [ ] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

### CODE REVIEW

File: app/client/src/sagas/PageSagas.tsx [hunk 1]
```diff
@@ -746,9 +746,29 @@ export function* clonePageSaga(
...
+          fetchJSCollectionsForPageSuccess([]).type,
+        ],
+        [
+          fetchActionsForPageError().type,
+          fetchJSCollectionsForPageError().type,
+        ],
+      );
+
+      if (!afterActionsFetch) {
+        throw new Error("Failed generating template");
```
Comment (ohansFavour) [inline app/client/src/sagas/PageSagas.tsx:767]: The saga here handles cloning of pages, this error message seems misleading as it suggests that a template is getting generated.

**Review by ohansFavour — COMMENTED** [review]
@Druthi Changes look good.

Can we add cypress tests here?., to catch potential regressions in the future

File: app/client/src/sagas/PageSagas.tsx (hunk 1 above)
Comment (Druthi) [reply app/client/src/sagas/PageSagas.tsx:767]: Oh yes. My bad. Changing it

Comment (Druthi) [reply app/client/src/sagas/PageSagas.tsx:767]: Im not sure how to add the test here as the error is cleared immediately on the second eval cycle.

File: app/client/cypress/integration/Regression_TestSuite/ClientSideTests/BugTests/Bug20841_Spec.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      paste: true,
+      completeReplace: true,
+      toRun: false,
+      shouldCreateNewJSObj: true,
+      prettify: false,
+    });
+
+    _.ee.DragDropWidgetNVerify(WIDGET.INPUT_V2, 200, 600);
+    _.ee.SelectEntityByName("Input1");
+    _.propPane.TypeTextIntoField("Default Value", "{{JSObject1.myFun1()}}");
```
Comment (Aishwarya-U-R) [inline app/client/cypress/integration/Regression_TestSuite/ClientSideTests/BugTests/Bug20841_Spec.ts:22]: UpdatePropertyFieldValue() can be used instead.

File: app/client/cypress/support/Pages/EntityExplorer.ts [hunk 3]
```diff
@@ -189,9 +185,10 @@ export class EntityExplorer {
     cy.get(this.locator._widgetInCanvas(widgetType)).should("exist");
   }
 
-  public ClonePage(pageName = "Page1") {
+  public ClonePage(pageName = "Page1", index = 0) {
```
Comment (Aishwarya-U-R) [inline app/client/cypress/support/Pages/EntityExplorer.ts:188]: Is adding index needed here?

File: app/client/cypress/integration/Regression_TestSuite/ClientSideTests/BugTests/Bug20841_Spec.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+    _.agHelper.AssertText(_.locators._inputWidget, "val", "Default text");
+
+    _.ee.ClonePage("Page1", 0);
+
+    _.agHelper.AssertText(_.locators._inputWidget, "val", "Default text");
+
+    _.ee.ClonePage("Page1", 0);
+
+    _.agHelper.AssertText(_.locators._inputWidget, "val", "Default text");
```
Comment (Aishwarya-U-R) [inline app/client/cypress/integration/Regression_TestSuite/ClientSideTests/BugTests/Bug20841_Spec.ts:32]: Clone again the Page1 Copy & verify the same.

File: app/client/cypress/support/Pages/EntityExplorer.ts (hunk 3 above)
Comment (Druthi) [reply app/client/cypress/support/Pages/EntityExplorer.ts:188]: Yes, it was erroring out cause there were two elements now since we already cloned a page

Comment (Aishwarya-U-R) [reply app/client/cypress/support/Pages/EntityExplorer.ts]: Updated method

**Review by Aishwarya-U-R — APPROVED** [review]

File: app/client/cypress/integration/Regression_TestSuite/ServerSideTests/Postgres_DataTypes/Binary_Spec.ts [hunk 5]
```diff
@@ -340,8 +340,8 @@ describe("Binary Datatype tests", function() {
     agHelper.ClickButton("DeleteQuery", 1);
     agHelper.ValidateNetworkStatus("@postExecute", 200);
     agHelper.ValidateNetworkStatus("@postExecute", 200);
-    agHelper.AssertElementAbsence(locator._spinner, 20000); //Allwowing time for delete to be success
-    agHelper.Sleep(4000); //Allwowing time for delete to be success
+    agHelper.AssertElementAbsence(locator._spinner, 20000); //Allowing time for delete to be success
+    agHelper.Sleep(6000); //Allwowing time for delete to be success
```
Comment (ohansFavour) [inline app/client/cypress/integration/Regression_TestSuite/ServerSideTests/Postgres_DataTypes/Binary_Spec.ts:344]: Did we regress on performance? Why are we increasing wait time here?

**Review by ohansFavour — APPROVED** [review]
The changes look good. Kindly address the performance concern @Druthi .

File: app/client/cypress/integration/Regression_TestSuite/ServerSideTests/Postgres_DataTypes/Binary_Spec.ts (hunk 5 above)
Comment (Aishwarya-U-R) [reply app/client/cypress/integration/Regression_TestSuite/ServerSideTests/Postgres_DataTypes/Binary_Spec.ts:344]: Nope, Binary_spec is a known to be bit flaky with respect to wait time already.. Hence increased time in this PR.

**Review by Aishwarya-U-R — APPROVED** [review]
Cypress looks good.

**Review by ohansFavour — APPROVED** [review]

**Pull request merged**